src/tests/testETag.cc

```diff
@@ -0,0 +1,201 @@
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "compat/cppunit.h"
+#include "ETag.h"
+#include "unitTestMain.h"
+
+#include <cstring>
+
+/*
+ * Unit tests for the ETag class which handles HTTP entity tag parsing
+ * and comparison as defined in RFC 7232.
+ */
+
+class TestETag : public CPPUNIT_NS::TestFixture
+{
+    CPPUNIT_TEST_SUITE(TestETag);
+    CPPUNIT_TEST(testParseStrongETag);
+    CPPUNIT_TEST(testParseWeakETag);
+    CPPUNIT_TEST(testParseInvalidETag);
+    CPPUNIT_TEST(testStrongEqualityMatching);
+    CPPUNIT_TEST(testStrongEqualityNonMatching);
+    CPPUNIT_TEST(testStrongEqualityWithWeakTags);
+    CPPUNIT_TEST(testWeakEqualityMatching);
+    CPPUNIT_TEST(testWeakEqualityNonMatching);
+    CPPUNIT_TEST(testWeakEqualityMixedStrength);
+    CPPUNIT_TEST_SUITE_END();
+
+protected:
+    void testParseStrongETag();
+    void testParseWeakETag();
+    void testParseInvalidETag();
+    void testStrongEqualityMatching();
+    void testStrongEqualityNonMatching();
+    void testStrongEqualityWithWeakTags();
+    void testWeakEqualityMatching();
+    void testWeakEqualityNonMatching();
+    void testWeakEqualityMixedStrength();
+};
+
+CPPUNIT_TEST_SUITE_REGISTRATION(TestETag);
+
+void
+TestETag::testParseStrongETag()
+{
+    ETag etag;
+    const char *strongTag = "\"strong-etag-value\"";
+    
+    auto result = etagParseInit(&etag, strongTag);
+    
+    CPPUNIT_ASSERT_EQUAL(1, result);
+    CPPUNIT_ASSERT(etag.str != nullptr);
+    CPPUNIT_ASSERT_EQUAL(0, etag.weak);
+    CPPUNIT_ASSERT_EQUAL(0, strcmp(etag.str, strongTag));
+}
+
+void
+TestETag::testParseWeakETag()
+{
+    ETag etag;
+    const char *weakTag = "W/\"weak-etag-value\"";
+    
+    auto result = etagParseInit(&etag, weakTag);
+    
+    CPPUNIT_ASSERT_EQUAL(1, result);
+    CPPUNIT_ASSERT(etag.str != nullptr);
+    CPPUNIT_ASSERT_EQUAL(1, etag.weak);
+    // str should point to the quoted part after "W/"
+    CPPUNIT_ASSERT_EQUAL(0, strcmp(etag.str, "\"weak-etag-value\""));
+}
+
+void
+TestETag::testParseInvalidETag()
+{
+    ETag etag;
+    
+    // Missing quotes
+    int result1 = etagParseInit(&etag, "no-quotes");
+    CPPUNIT_ASSERT_EQUAL(0, result1);
+    CPPUNIT_ASSERT(etag.str == nullptr);
+    
+    // Only opening quote
+    int result2 = etagParseInit(&etag, "\"missing-end");
+    CPPUNIT_ASSERT_EQUAL(0, result2);
+    CPPUNIT_ASSERT(etag.str == nullptr);
+    
+    // Only closing quote
+    int result3 = etagParseInit(&etag, "missing-start\"");
+    CPPUNIT_ASSERT_EQUAL(0, result3);
+    CPPUNIT_ASSERT(etag.str == nullptr);
+    
+    // Empty string
+    int result4 = etagParseInit(&etag, "");
+    CPPUNIT_ASSERT_EQUAL(0, result4);
+    CPPUNIT_ASSERT(etag.str == nullptr);
+    
+    // Single quote
+    int result5 = etagParseInit(&etag, "\"");
+    CPPUNIT_ASSERT_EQUAL(0, result5);
+    CPPUNIT_ASSERT(etag.str == nullptr);
+}
+
+void
+TestETag::testStrongEqualityMatching()
```

Comment:
It would be good to fold these all the functions testing `etagIsStrongEqual()` into this one.